source/common/router/config_impl.cc

```diff
@@ -108,10 +108,17 @@ RouteEntryImplBaseConstSharedPtr createAndValidateRoute(
   if (validation_clusters.has_value()) {
     route->validateClusters(*validation_clusters);
     for (const auto& shadow_policy : route->shadowPolicies()) {
-      ASSERT(!shadow_policy->cluster().empty());
-      if (!validation_clusters->hasCluster(shadow_policy->cluster())) {
-        throw EnvoyException(
-            fmt::format("route: unknown shadow cluster '{}'", shadow_policy->cluster()));
+      if (!shadow_policy->cluster().empty()) {
+        ASSERT(shadow_policy->clusterHeader().get().empty());
+        if (!validation_clusters->hasCluster(shadow_policy->cluster())) {
+          throw EnvoyException(
+              fmt::format("route: unknown shadow cluster '{}'", shadow_policy->cluster()));
+        }
+      } else {
+        // For shadow policies with `cluster_header_`, we only verify that this field is not
+        // empty because the cluster name is not set yet at config time.
+        // Assert here after the validateMirrorClusterSpecifier check.
+        ASSERT(!shadow_policy->clusterHeader().get().empty());
```

Comment:
Because we put the `validateMirrorClusterSpecifier` in the ctor of `ShdowPolicyImpl` and we throw the exception below, we can remove this assertion. There are unit tests to test the failure cases. This is to verify route after the shadow policy is set, and that's why I put assertion here. We can remove it I think.